base/meta.jl

```diff
@@ -7,17 +7,66 @@ module Meta
 
 export quot,
        isexpr,
+       propagate_sourceloc!,
        show_sexpr
 
+"""
+    quot(ex)
+
+Return `Expr(:quote, ex)`.  This is useful for interpolating symbols into an
+AST as symbol literals rather than variable names.
+"""
 quot(ex) = Expr(:quote, ex)
 
+"""
+    isexpr(ex, head [, n])
+
+Return true if `ex` is an `Expr` with the given type `head` and optionally that
+the argument list is of length `n`.  `head` may be a `Symbol` or collection of
+`Symbol`s.  For example, to check that a macro was passed a function call
+expression, you might use `isexpr(ex, :call)`.
+"""
 isexpr(ex::Expr, head)          = ex.head === head
 isexpr(ex::Expr, heads::Union{Set,Vector,Tuple}) = in(ex.head, heads)
 isexpr(ex,       head)          = false
 
 isexpr(ex,       head, n::Int)  = isexpr(ex, head) && length(ex.args) == n
 
 
+"""
+    propagate_sourceloc!(location, expr)
+
+Overwrite the caller source location embedded in the macro call AST for all
+macro calls inside `expr`, and return the resulting AST.  This is most useful
+to make the `@__LINE__` and `@__FILE__` macros return the location where a
+parent macro was invoked.  For example:
+
+```
+macro example()
+    propagate_sourceloc!(__source__,
+        quote
+            @__LINE__
```

Comment:
This is somewhat related to that old comment, but addresses the more general problem of wrapping macros which make use of `__source__`, for example here:

https://github.com/c42f/MicroLogging.jl/blob/e37dffcb7aeb9c581a27b897fa9d4fe3467d8b85/src/MicroLogging.jl#L205

If it were just `@__LINE__` I was worried about, interpolating `__source__.line` would be a better solution.